File: 04-stack/00739-Daily_Temperatures/00739-daily_temperatures.go
// 739: Daily Temperstures
// https://leetcode.com/problems/daily-temperatures/

package main

import "fmt"

// SOLUTION
func dailyTemperatures(temperatures []int) []int {
	var s []int
	result := make([]int, len(temperatures))

	for i:=0; i<len(temperatures); i++ {
		for len(s)!=0 && temperatures[i]>temperatures[s[len(s)-1]] {
			result[s[len(s)-1]] = i - s[len(s)-1]
			s = s[:len(s)-1]
		}
		s = append(s, i);
	}

	return result;
}

func main() {
	// INPUT
	temperatures := []int{73,74,75,71,69,72,76,73}

	// OUTPUT
	result := dailyTemperatures(temperatures)
	fmt.Println(result)
}

File: 04-stack/00739-Daily_Temperatures/00739-daily_temperatures.java
// 739: Daily Temperstures
// https://leetcode.com/problems/daily-temperatures/


import java.util.Stack;

class Solution {
    // SOLUTION
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> s = new Stack<>();
        int[] result = new int[temperatures.length];

        for (int i=0; i<temperatures.length; i++) {
            while (!s.isEmpty() && temperatures[i]>temperatures[s.peek()]) {
                result[s.peek()] = i - s.peek();
                s.pop();
            }
            s.push(i);
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] temperatures = {73,74,75,71,69,72,76,73};

        // OUTPUT
        var result = o.dailyTemperatures(temperatures);
        System.out.print("["); for (var v : result) System.out.print(v+" "); System.out.println("\b]");
    }
}